### PR TITLE
Can ignore pulse_number column on TOA read/write

### PR DESCRIPTION
**Pull request by dlakaplan**
Reading or writing TOAs with `include_pn=False` will ignore the `pulse_number` column. It will issue a warning that this is happening.  This may make it easier to merge TOA files, but in general it may also be dangerous.  So I'm not convinced that this is overall a good idea but it will address #1397 

The error message when failing to merge is also a little more informative.

### CODE REVIEW

**Comment by dlakaplan** [comment]
#1400 may also be relevant.

**Comment by dlakaplan** [comment]
Example error trace (now):
```
WARNING  (pint.toa                      ): File 5 has extra column(s): pulse_number
Traceback (most recent call last):
  File "/Users/kaplan/pythonpackages/PINT_dlakaplan/./test_merge.py", line 28, in <module>
    tall = pint.toa.merge_TOAs(t)
  File "/Users/kaplan/pythonpackages/PINT_dlakaplan/src/pint/toa.py", line 2451, in merge_TOAs
    raise TypeError(
TypeError: merge_TOAs() cannot merge. Inconsistent numbers of table columns: [13, 13, 13, 13, 13, 14, 13]
```
is that reasonable?

**Comment by scottransom** [comment]
I think that that TypeError looks pretty good.

**Comment by aarchiba** [comment]
> I think that that TypeError looks pretty good.

Isn't that basically the error we were getting? That mentions inconsistent numbers of columns but gives no hint which columns are missing/extra/

**Comment by aarchiba** [comment]
Perhaps a convenience function to drop the pulse number column if present would be a more natural way to address the problem of spurious pulse number columns? Plus, more directly, merge_TOAs can know about pulse number columns and add one that's all NaNs if necessary.

**Comment by scottransom** [comment]
I personally don't even really think this is an "issue".  The real issue is that NANOGrav shouldn't have been writing pulse numbers into our exported TOA files.  That is putting model specific things in a .tim file (although hopefully they are *relatively* correct between proper models, at least).  I think I'd worry about fake pulse number columns with NaNs possibly even causing more problems.

**Comment by aarchiba** [comment]
> I personally don't even really think this is an "issue". The real issue is that NANOGrav shouldn't have been writing pulse numbers into our exported TOA files. That is putting model specific things in a .tim file (although hopefully they are _relatively_ correct between proper models, at least). I think I'd worry about fake pulse number columns with NaNs possibly even causing more problems.

We have the limitation currently that unless all TOAs have pulse numbers, pulse numbers are ignored. So it won't cause any problems. But we should fix that (tied up in the way we compute absolute phases) for other reasons anyway. There are many situations where the pulse numbers can be regarded as known while we vary our models, including all the triple-system work, so it's perfectly reasonable to include pulse numbers when passing TOAs to a different optimizer.

The problem we do have - problem in the sense that our users are stuck and frustrated - is that they can't merge perfectly reasonable sets of TOAs and we don't tell them enough information for them to resolve the issue.

**Comment by dlakaplan** [comment]
I am not sure of the best approach.  Right now this PR will
* provide warnings of incompatible numbers/names of columns (although the actual exception will still be the same), so users will hopefully understand better what is happening
* allow users to ignore pulse numbers on read and/or write, which can solve any incompatibility 

It does not include any NaN or other info for a dummy pulse number column, although that's easy.  But I didn't see any documentation of that.  if we decide that this is a better way to go I can add it (together with a warning) easily.

File: tests/test_toa_reader.py [hunk 1]
```diff
@@ -234,6 +235,51 @@ def test_toa_merge_different_ephem():
...
+def test_toa_merge_different_columns():
+    filenames = [
+        datadir / "NGC6440E.tim",
+        datadir / "testtimes.tim",
+    ]
+    model = get_model(datadir / "NGC6440E.par")
+    # add a pulse_number column.  then the merge should fail
+    toas = [toa.get_TOAs(ff, model=model) for ff in filenames]
+    toas[0].compute_pulse_numbers(model)
+    with pytest.raises(TypeError):
```
Comment (aarchiba) [inline tests/test_toa_reader.py:247]: Can you check that the exception message contains the name of the offending column? Pytest provides some tools for asserting about exception messages.

**Comment by aarchiba** [comment]
I don't like adding even more arguments to our reading and writing functions. I'd prefer a function to drop pulse numbers (if present), right next to the function to add pulse numbers (based on a model). This achieves the same effect.

**Comment by dlakaplan** [comment]
Now it will not write `pn=nan` to the tim file.  a more general solution may be useful, but it's hard since flags are strings.

File: src/pint/toa.py [hunk 2]
```diff
@@ -305,6 +308,9 @@ def get_TOAs(
     if usepickle and updatepickle:
         log.info("Pickling TOAs.")
         save_pickle(t, picklefilename=picklefilename)
+    if "pulse_number" in t.table.colnames and not include_pn:
+        log.warning(f"'pulse_number' column exists but not being read in")
```
Comment (aarchiba) [inline src/pint/toa.py:312]: If the user has explicitly asked to ignore pulse numbers I don't think they should get a warning if pulse numbers are encountered. Nothing at all, or at most an info/debug-level note.

File: src/pint/toa.py [hunk 3]
```diff
@@ -305,6 +308,9 @@ def get_TOAs(
     if usepickle and updatepickle:
         log.info("Pickling TOAs.")
         save_pickle(t, picklefilename=picklefilename)
+    if "pulse_number" in t.table.colnames and not include_pn:
```
Comment (aarchiba) [inline src/pint/toa.py:311]: Checking `include_pn` first makes it more obvious that this is controlled by an option, and also saves doing the `in` in the common (default) case where the option isn't set.

File: src/pint/toa.py [hunk 4]
```diff
@@ -1888,6 +1895,16 @@ def compute_pulse_numbers(self, model):
         self.table["pulse_number"] = phases.int
         self.table["pulse_number"].unit = u.dimensionless_unscaled
 
+    def remove_pulse_numbers(self):
+        """Delete pulse numbers from TOA data"""
+
+        if "pulse_number" in self.table.colnames:
+            del self.table["pulse_number"]
+        else:
+            log.warning(
```
Comment (aarchiba) [inline src/pint/toa.py:1904]: I'd rather see this as "delete pulse numbers if present" and have it run smoothly with no warnings if they're not present - then this is a cheap way to make sure there are no pulse numbers (just before saving, say).

File: src/pint/toa.py [hunk 5]
```diff
@@ -1951,6 +1968,7 @@ def write_TOA_file(
         commentflag=None,
         order_by_index=True,
         *,
+        include_pn=True,
```
Comment (aarchiba) [inline src/pint/toa.py:1971]: I think it's easier to just `remove_pulse_numbers` before saving rather than add yet another option to `write_TOA_file`.

File: src/pint/toa.py [hunk 6]
```diff
@@ -2002,7 +2022,15 @@ def write_TOA_file(
         # FIXME: everywhere else the pulse number column is called pulse_number not pn
         toacopy = copy.deepcopy(self)
         if "pulse_number" in toacopy.table.colnames:
-            toacopy["pn"] = toacopy.table["pulse_number"]
+            if include_pn:
+                toacopy["pn"] = toacopy.table["pulse_number"]
+                # but delete those that are NaN
+                for i in np.where(np.isnan(self["pulse_number"]))[0]:
+                    del toacopy.table["flags"][i]["pn"]
+            else:
```
Comment (aarchiba) [inline src/pint/toa.py:2030]: Again, no warning about something explicitly requested.

File: src/pint/toa.py [hunk 7]
```diff
@@ -2420,11 +2453,66 @@ def merge_TOAs(TOAs_list):
...
+    # check for the presence of various computable columns
+    #   pulse_number
+    #   ssb_obs_pos, ssb_obs_vel, obs_sun_pos
+    #   tdb, tdbld
+    # if one file has these and the others don't, try to add them if strict=False
+    all_colnames = [tt.table.colnames for tt in TOAs_list]
+    has_pulse_number = np.array(
+        ["pulse_number" in colnames for colnames in all_colnames]
+    )
+    has_posvel = np.array(
```
Comment (aarchiba) [inline src/pint/toa.py:2465]: This logic isn't quite right if some have a few columns but not all. (For example, some have ssb_obs_pos but not ssb_obs_vel.

File: src/pint/toa.py [hunk 8]
```diff
@@ -2420,11 +2453,66 @@ def merge_TOAs(TOAs_list):
...
+            log.warning("Not all data have 'pulse_number' columns but strict=True")
+        else:
+            # some data have pulse_numbers but not all
+            # put in NaN
+            for i, tt in enumerate(TOAs_list):
+                if not "pulse_number" in tt.table.colnames:
+                    log.warning(
+                        f"'pulse_number' not present in data set {i}: inserting NaNs"
+                    )
+                    tt.table.add_column(np.ones(len(tt)) * np.nan, name="pulse_number")
```
Comment (aarchiba) [inline src/pint/toa.py:2490]: Doesn't this modify the input table? I don't think `merge_TOAs` should ever modify its inputs.

File: src/pint/toa.py [hunk 9]
```diff
@@ -2420,11 +2453,66 @@ def merge_TOAs(TOAs_list):
...
+            for colnames in all_colnames
+        ]
+    )
+    has_tdb = np.array(
+        [("tdb" in colnames) and ("tdbld" in colnames) for colnames in all_colnames]
+    )
+    has_posvel_ecl = np.array(
+        ["ssb_obs_vel_ecl" in colnames for colnames in all_colnames]
+    )
+    if has_pulse_number.any() and not has_pulse_number.all():
```
Comment (aarchiba) [inline src/pint/toa.py:2479]: You don't need the all() check, any that has the required columns will be untouched.

File: tests/test_pulse_number.py [hunk 10]
```diff
@@ -64,6 +64,12 @@ def test_pulse_number(model, toas):
     assert "pulse_number" in toas.table.colnames
 
 
+def test_remove_pulse_number(model, toas):
```
Comment (aarchiba) [inline tests/test_pulse_number.py:67]: It'd be good to have a test to demonstrate what happens when pulse numbers aren't present and are removed - this is where one specifies whether or not there should be a warning.

File: tests/test_toa_reader.py [hunk 11]
```diff
@@ -234,6 +236,121 @@ def test_toa_merge_different_ephem():
...
+    t = toa.get_TOAs(filenames[0], model=model)
+    t.compute_pulse_numbers(model)
+    f = StringIO()
+    t.write_TOA_file(f)
+    f.seek(0)
+    toas = [toa.get_TOAs(fname, include_pn=False) for fname in [f, filenames[1]]]
+    nt = toa.merge_TOAs(toas)
+
+
+def test_toa_merge_different_columns_ignorepn_onwrite():
```
Comment (aarchiba) [inline tests/test_toa_reader.py:338]: Is there any need to test the combination of merging and `ignore_pn`? They seem to be unrelated features, easily tested separately.

File: src/pint/toa.py [hunk 12]
```diff
@@ -103,6 +103,7 @@ def get_TOAs(
     bipm_version=None,
     include_gps=None,
     planets=None,
+    include_pn=True,
```
Comment (aarchiba) [inline src/pint/toa.py:106]: Is there any need for this? The column can easily be removed with `remove_pulse_numbers`.

File: tests/test_toa_writer.py [hunk 13]
```diff
@@ -118,6 +118,23 @@ def test_roundtrip_ncyobs_toa_TEMPOformat(tmp_path):
...
+    model = get_model(datadir / "NGC6440E.par")
+    # read and add pulse numbers
+    t = toa.get_TOAs(datadir / "NGC6440E.tim", model=model)
+    t.compute_pulse_numbers(model)
+    t["pulse_number"][10] = np.nan
+    f = StringIO()
+    t.write_TOA_file(f)
+    f.seek(0)
+    contents = f.read()
+    assert "nan" not in contents
```
Comment (aarchiba) [inline tests/test_toa_writer.py:132]: Best to convert contents to lowercase first to rule out writing NaN as well.

File: src/pint/toa.py (hunk 12 above)
Comment (aarchiba) [inline src/pint/toa.py:106]: Is there any need for this? The column can easily be removed with `remove_pulse_numbers`.

File: tests/test_toa_writer.py (hunk 13 above)
Comment (aarchiba) [inline tests/test_toa_writer.py:132]: Best to convert contents to lowercase first to rule out writing NaN as well.

File: src/pint/toa.py (hunk 3 above)
Comment (dlakaplan) [reply src/pint/toa.py:311]: ok

File: src/pint/toa.py (hunk 2 above)
Comment (dlakaplan) [reply src/pint/toa.py:312]: ok, changed to debug

File: src/pint/toa.py (hunk 4 above)
Comment (dlakaplan) [reply src/pint/toa.py:1904]: ok

File: src/pint/toa.py (hunk 6 above)
Comment (dlakaplan) [reply src/pint/toa.py:2030]: ok

File: src/pint/toa.py [hunk 14]
```diff
@@ -2420,11 +2453,66 @@ def merge_TOAs(TOAs_list):
...
+    )
+    has_tdb = np.array(
+        [("tdb" in colnames) and ("tdbld" in colnames) for colnames in all_colnames]
+    )
+    has_posvel_ecl = np.array(
+        ["ssb_obs_vel_ecl" in colnames for colnames in all_colnames]
+    )
+    if has_pulse_number.any() and not has_pulse_number.all():
+        if strict:
+            log.warning("Not all data have 'pulse_number' columns but strict=True")
```
[the remark this answers is not in the document]
Comment (dlakaplan) [reply src/pint/toa.py:2481]: ok

File: src/pint/toa.py (hunk 9 above)
Comment (dlakaplan) [reply src/pint/toa.py:2479]: but the obsvel and TDB functions remove previous columns, so I thought that would take unnecessary time.

File: tests/test_pulse_number.py (hunk 10 above)
Comment (dlakaplan) [reply tests/test_pulse_number.py:67]: ok, added

File: tests/test_toa_writer.py (hunk 13 above)
Comment (dlakaplan) [reply tests/test_toa_writer.py:132]: good point

File: src/pint/toa.py (hunk 12 above)
Comment (dlakaplan) [reply src/pint/toa.py:106]: especially for reading, I think it's needed.  this directly addresses the use case Deborah had where multiple files are being read in together so you can't remove pulse numbers individually.

File: tests/test_toa_reader.py (hunk 11 above)
Comment (dlakaplan) [reply tests/test_toa_reader.py:338]: I guess not.

File: src/pint/toa.py (hunk 12 above)
Comment (scottransom) [reply src/pint/toa.py:106]: I think I'm with @aarchiba on this one.  That seems like a really specific and unneeded option.  Because even if you read the files together, you still need to merge them, and so can remove pulse numbers before you do that.  Unless I'm missing something...

Comment (dlakaplan) [reply src/pint/toa.py:106]: Merging is handled within `get_TOAs()`.  I guess with this now inserting pulse numbers by default then it's fine to remove this option, although it may be that instead the `strict` option should be used instead.

Comment (dlakaplan) [reply src/pint/toa.py:106]: Maybe a `strict` option isn't needed, but I personally prefer to be able to control if the data input will be additionally modified on read.
